### ISSUE TITLE
Indicators discussion

### ISSUE THREAD

**Issue by wlthomson** [openmsupply/mobile#1685]
## Description

Epic for organising indicators work for `v3.2.0` release.

### Context

HIV indicators.

## Design discussion

As the desktop indicators implementation is still awaiting some feedback (in particular re. migration and sync) there are still some unknowns which make it hard to make informed back-end design decisions. If others agree, it seems like a good idea to focus discussion on front-end for now, just so that we can keep mobile and desktop development on indicators going in parallel.

Initial thoughts on general design/implementation process:

### Front-end

- [x] UI design.
   - [x] Wireframe e.g. prototypes.
   - [x] Evaluate designs (time constraints, performance, UX etc.).
   - [x] Wireframe final modal/page designs.
- [x] Components
  - [x] Decide what existing/new components are needed to implement chosen design.
  - [x] Implement any new components.
- [x] Navigation
  - [x] Decide navigation logic between any new pages, modals. 
  - [x] Implement navigation logic.

### Back-end

- [x] Database
  - [x] Decide what data is needed from desktop.
  - [x] Design indicators schema.
  - [x] Implement new tables, fields.
- [x] Sync
  - [x] Implement incoming sync logic.
  - [x] Implement version migration logic (including messages?).
  - [x] Implement outgoing sync logic.
- [x] Logic
  - [x] Implement indicators toggle.
  - [x] Implement indicators dropdown.
  - [x] Implement indicators dynamic columns.
  - [x] Implement indicator values rendering to data table.
  - [x] Implement indicator values cell editing.

### Tab (items)

![prototype-tab-items](https://user-images.githubusercontent.com/43223637/71567738-32d82880-2b26-11ea-811e-2e7038cecf90.png)

### Tab (indicators, no selection)

![prototype-tab-indicators-no-select](https://user-images.githubusercontent.com/43223637/71567817-be51b980-2b26-11ea-9b51-1d61c61b2ab6.png)

### Tab (indicators, selection, not expanded)

![prototype-tab-indicators-select-no-expand](https://user-images.githubusercontent.com/43223637/71567802-a2e6ae80-2b26-11ea-979b-9cf011693be2.png)

### Tab (indicators, selection, expanded)

![prototype-tab-indicators-select-expand](https://user-images.githubusercontent.com/43223637/71567804-a5490880-2b26-11ea-849d-df8ffe85efe0.png)

### Page (no selection). 

![prototype-page-indicators-no-select](https://user-images.githubusercontent.com/43223637/71567744-39ff3680-2b26-11ea-86b0-90cc1631fdcb.png)

### Page (selection).

![prototype-page-indicators-select](https://user-images.githubusercontent.com/43223637/71567788-7cc10e80-2b26-11ea-8442-a7815d46345d.png)

### Stepper.

![prototype-stepper](https://user-images.githubusercontent.com/43223637/71567789-8185c280-2b26-11ea-8735-cc06fc87fac9.png)

## Issues.

- #1823.
- #1828.
- #1829.
- #1830.
- #1831.
- #1842.
- #1852.
- #1877.
- #1901.
- #1902.
- #1903.
- #1906.
- #1913.
- #1915.
- #1925.

**Comment by joshxg** [comment]
After getting a little more familiar with desktop seems the interface might be a little more annoying than anticipated! 

It is clear to me at least, that there needs to be some form of 'secondary page' - at the moment this is the regimen modal. Whether it be tab/stepper (as you've suggested above) or a modal, the 'secondary page' also needs to essentially have infinitely many different table configurations, which is annoying!

I think a stepper is possibly too much work to be done in this time frame and a modal is not-ideal.

My 2c is to use tabs on the `SupplierRequisitionPage`, to switch between `Items` and `Regimen/Indicators/whatever called` - then the `Regimen` tab has a drop down menu to select which set of indicators to display. Although thinking about it more, it might also be too much work, as each "tab" will not *just* be the table, as the buttons "Add from masterlist" etc would then need to change/switch out ? Although this would imo be really cool for stocktakes too!

**Comment by wlthomson** [comment]
Yeah, the main problem, as you say, is using a design pattern that allows for potentially unlimited indicator "groupings". _However_, in practice, each program is probably not going to have more than a few indicators. Tbh, once the existing regimen data UI is properly migrated, the average program will probably only have one indicator set (e.g. a HIV program may collect data on ART treatments for children, adults etc., a malaria program may collect data on anti-malarials for men, women etc. etc.). The only time you'll need multiple sets is when different data (i.e. attributes, i.e. columns) need to be collected for different regimes/regimens (e.g. some regimes want to collect separate data for men and women, some for children and adults). So far everything I've seen has indicated that this is the exception rather than the norm. 

With that in mind, my initial thoughts for the IC deliverables is:

1. we should prefer UI designs which can accommodate multiple indicators.
2. if there is a trade-off, a good UI design which only supports a small number of indicators, e.g. ~10, is preferable to a less usable design which supports a very number of indicators (one program having a number of indicators in the double digits seems impractical from a data entry standpoint, regardless of UI).
3. design the implementation to be extendable, but focus on depth rather than breadth, i.e. prioritise getting a good implementation working for regimen and HIV indicators. 

The idea I had in mind was quite similar to what you have described, but I was thinking of whether we could navigate to a new page for the indicator data? That way, we could just pass it props/dispatch in the usual way and the dynamic component switching wouldn't be needed?

Basic idea:

- Add tab/navigation arrow to `SupplierRequisitionPage` to switch between `Items` and `Indicators`.
  - For desktop the items tab is called `Data entry`, but IMO this is overly generic and could refer to indicators as well as items. 
  - I prefer `Indicators` over `Regimens` just because we are currently using the term indicator as a generalisation of regimen (i.e. a regimen is a specific type of indicator). From what I've read/been told, I think that is incorrect and that regimens is actually the right term, but for now I think do it how IC want it. 
- The tab/arrow brings up a new view/page containing one or two scrollable tables (aligned vertically). 
  - Basically the same as how it is done in desktop, so a safe choice for UAT (as I understand from brief chat with @gary, IC passed the current indicators UI without any issues).
  - Can be relatively easily extended to dynamically "switch out" tables using a dropdown or other selection widget.

Just finishing some indicators desktop fixes. Will add wireframes soon to give a better idea. I think a good target is to have agreement on the design locked down by the end of the week (ideally earlier, but being cautious re. desktop indicator workload).

**Comment by joshxg** [comment]
I think we're on the same page. Lets do the wireframes you describe and go from there. 

I'd just suggest getting the wireframes done earlier rather than later as there will be some back and forth. Keep em as simple structures and they'll only take a few minutes to iterate on

**Comment by joshxg** [comment]
@wlthomson 

Personally I like the first wireframe and the selection. 

- My reasoning for liking the tabs over the page is that it's already sort of two levels deep from the plural page -> non-plural page .. I think going deeper just doesn't sort of feel right? Although it is more of a switch navigation rather than a stack that you're suggesting. I think both could probably work, but I'm leaning more toward the tabs!
- I like the selection because of the keyboard. I just don't think there is enough real estate to accommodate two tables. The keyboard is going to cover the bottom one (and will be annoying to code to do that!) and get really annoying to be able to switch between em

**Comment by wlthomson** [comment]
> @wlthomson
> 
> Personally I like the first wireframe and the selection.
> 
>     * My reasoning for liking the tabs over the page is that it's already sort of two levels deep from the plural page -> non-plural page .. I think going deeper just doesn't sort of feel right? Although it is more of a switch navigation rather than a stack that you're suggesting. I think both could probably work, but I'm leaning more toward the tabs!
> 
>     * I like the selection because of the keyboard. I just don't think there is enough real estate to accommodate two tables. The keyboard is going to cover the bottom one (and will be annoying to code to do that!) and get really annoying to be able to switch between em
> 
> 
> Also: Could potentially shorten the search bar and put the toggle on the LHS of that. It's pretty long. Although I don't think that gives much more vertical space, if any..

I agree.

I think shortening the search bar could work, but I like the idea of the LHS being invariant and only the RHS updating when toggled. What do you think?

Have added my idea of finalised wireframes (see tabs with selection).

**Comment by joshxg** [comment]
I like the tabs one even more now! nice

**Comment by joshxg** [comment]
@wlthomson 

I have made a drop down menu component on the dispensing feature branch. I will do a PR for it in develop so that you can use it, if you'd like?

The rest of the components are all already in including the `ToggleBar` component. I've also done something similar for the `indicators/items` toggle on the data for `prescribers/patients` - essentially just using a new field in the redux store of the page, `dataSet`

**Comment by joshxg** [comment]
My thoughts currently:

### Front-end

- [x] UI design.
   - [x] Wireframe e.g. prototypes.
   - [x] Evaluate designs (time constraints, performance, UX etc.).
   - [x] Wireframe final modal/page designs.
- [x] Components - I don't see any new ones being needed
  - [x] Decide what existing/new components are needed to implement chosen design.
  - [x] Implement any new components.
- [x] Navigation - I don't see any navigation logic being needed
  - [x] Decide navigation logic between any new pages, modals. 
  - [x] Implement navigation logic.

### Back-end

- [ ] Database - I think doing this basically exactly the same as desktop? Obviously slightly different casing/types/defaults? An example of adding directions/abbreviations https://github.com/openmsupply/mobile/pull/1788/files
  - [ ] Decide what data is needed from desktop.
  - [ ] Design indicators schema.
  - [ ] Implement new tables, fields.
- [ ] Sync - Maybe I'm not seeing it but I don't see too much needed here, just pretty much mappings. Recent PR example: https://github.com/openmsupply/mobile/pull/1789
  - [ ] Design 4D -> Realm mapping.
  - [ ] Implement incoming sync logic.
  - [ ] Design Realm -> 4D mapping (what fields can be edited by mobile?)
  - [ ] Implement outgoing sync logic.
- [ ] State management - this is probs where most of the work is, but I don't see it being too heavy. Really just need some actions for managing editing of a cell?
  - [ ] Design state diagram for indicators.
  - [ ] Decompose state diagram into actions and reducers.
  - [ ] Implement actions and reducers.


Couple things not noted would be dynamic columns for the indicator tables (see: `getColumns.js`), possibly some helper functions for organizing a `program_indicator` and it's rows/columns into two arrays. This kind of blends with the above regarding columns, but basically mapping each column to a `column object` similar to that in `getColumns.js` and also each `row` into a row object similar to the ones currently used - `TransactionItem`. `TransactionBatch` etc. Really basic mapping functions

Also of course making the edits to the `SupplierRequisitionPage` for indicators. I really think it's unlikely we need to do `CustomerRequisitionPage` also.

**Comment by wlthomson** [comment]
Description roadmap updated with current progress.

Mobile logic is nearly done, just need to add cell logic, then a tidy up. Some table changes are also needed desktop side, but nothing major. 

As I see it, the main remaining issue is ensuring data integrity on version upgrade and related messages/migration logic (see related issue #1849). Getting the basic messages implementation going in mobile should be quite straightforward, but there are a few aspects of the version upgrade messaging "architecture" that I think still need to be discussed. 

@wlthomson @joshxg @andreievg, @mark-prins Are we free to have a discussion today to get the details for the version upgrade messaging agreed?

**Comment by joshxg** [comment]
@wlthomson I'm not in today so I won't be there. I was planning to do a messages PR today with tables and sync logic. I'll hold off if you are planning to have a meeting and just update the issue with what you've decided?

**Comment by wlthomson** [comment]
> @wlthomson I'm not in today so I won't be there. I was planning to do a messages PR today with tables and sync logic. I'll hold off if you are planning to have a meeting and just update the issue with what you've decided?

Just adding in case needed for future reference: see https://github.com/sussol/msupply/issues/5068 for messages update.

**Comment by joshxg** [comment]
#1928 has been merged. Going to leave this open until the docs and test cases are complete

**Comment by joshxg** [comment]
Docs have been written - just waiting on the final copy but going to close this now! Thanks @wlthomson
